crates/valence_server/src/placement.rs

```diff
@@ -0,0 +1,939 @@
+    MuddyMangroveRoots,
+    MushroomBlocks,
+    NetherWart,
+    NetherPortal,
+    NoteBlock,
+    Observer,
+    PinkPetals,
+    Pistons,
+    MovingPiston,
+    PistonHead,
+    Planks,
+    Potatoes,
+    PointedDripstone,
+    PressurePlates,
+    Prismarine,
+    PumpkinAndCarvedPumpkin,
+    PumpkinStem,
+    PurpurAndQuartzPillar,
+    Rails,
+    Rail,
+    ActivatorRailAndDetectorRailAndPoweredRail,
+    RedstoneComparator,
+    RedstoneDust,
+    RedstoneLamp,
+    RedstoneOre,
+    RedstoneRepeater,
+    RedstoneTorch,
+    RespawnAnchor,
+    //Bedrock Only: SandAndRedSand,
+    //Bedrock Only: SandstoneAndRedSandstone,
+    Saplings,
+    Scaffolding,
+    SculkCatalyst,
+    SculkSensor,
+    SculkShrieker,
+    SculkVein,
+    SeaPickle,
+    ShulkerBoxes,
+    Sign,
+    Slabs,
+    SmallDripleaf,
+    Smoker,
+    //Missing from the overview:
+    SnifferEgg,
+    Snow,
+    Sponge,
+    Stairs,
+    Stones,
+    StoneBricks,
+    Stonecutter,
+    StructureBlock,
+    StructureVoid,
+    SugarCane,
+    SweetBerryBush,
+    TallGrassAndLargeFern,
+    TallSeagrass,
+    Target,
+    Terracottas,
+    TNT,
+    TorchAndSoulTorch,
+    Trapdoors,
+    Tripwire,
+    TripwireHook,
+    TurtleEgg,
+    TwistingVines,
+    UnderwaterTorch,
+    Vines,
+    Walls,
+    WeepingVines,
+    WheatCrop,
+    Wood,
+    Water,
+    Lava,
+}
+
+#[derive(Clone)]
+pub enum CardinalDirection {
+    North,
+    West,
+    South,
+    East,
+}
+
+impl CardinalDirection {
+    pub const fn into_prop_value(self) -> PropValue {
+        match self {
+            Self::North => PropValue::North,
+            Self::West => PropValue::West,
+            Self::South => PropValue::South,
+            Self::East => PropValue::East,
+        }
+    }
+
+    pub const fn opposite(self) -> Self {
+        match self {
+            Self::North => Self::South,
+            Self::West => Self::East,
+            Self::South => Self::North,
+            Self::East => Self::West,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Half {
+    Bottom,
+    Top,
+}
+
+impl Half {
+    pub fn into_prop_value(self) -> PropValue {
+        match self {
+            Half::Bottom => PropValue::Bottom,
+            Half::Top => PropValue::Top,
+        }
+    }
+
+    pub const fn opposite(self) -> Self {
+        match self {
+            Self::Bottom => Self::Top,
+            Self::Top => Self::Bottom,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub enum Side {
+    Left,
+    Right,
+}
+
+impl Side {
+    pub fn into_prop_value(self) -> PropValue {
+        match self {
+            Side::Left => PropValue::Left,
+            Side::Right => PropValue::Right,
+        }
+    }
+
+    pub const fn opposite(self) -> Self {
+        match self {
+            Self::Left => Self::Right,
+            Self::Right => Self::Left,
+        }
+    }
+}
+
+impl PlacementRule {
+    //TODO: WIP
+    pub const fn from_kind(kind: BlockKind) -> Self {
+        match kind {
```

Comment:
Unfortunately this approach isn't really maintainable. I don't want to manually update this list every time a new Minecraft version is released.

I think we should investigate how Vanilla is mapping the blocks to their placement rules and extract that information if possible. If not, then perhaps we could use some heuristic based on properties, but this is not my desired solution.

Reply:
I don't think there's much room for alternatives, if we want to mimic the client behaviour. Minecraft seems to hardcode these rules for these sets of blocks as well. As you said these rules are very complicated and I don't see a good way around this. Having a good default, recognizing similarities and thus being able to deduplicate code are things I try to do and hopefully can limit the maintanence cost. Also adding new blocks to these rules shouldn't be too much work.

Only alternative I can think of is dropping this in the scope of Valence entirely and make a crate for vanilla behaviour, which is allowed to be behind Valence and the newest Minecraft update a bit.

Reply:
Yeah minecraft implements placement behavior for each block by implementing this function:

```java
public BlockState getPlacementState(ItemPlacementContext ctx)
```

For example, this is the impl in `EnderChestBlock`:

```java
public BlockState getPlacementState(ItemPlacementContext ctx) {
        FluidState fluidState = ctx.getWorld().getFluidState(ctx.getBlockPos());
        return (BlockState)((BlockState)this.getDefaultState().with(FACING, ctx.getHorizontalPlayerFacing().getOpposite())).with(WATERLOGGED, fluidState.getFluid() == Fluids.WATER);
    }
```

It might be feasible to have some codegen that reads the `with()` calls and derives the correct placement rule for the block. eg, if `with(FACING, ...)` is present, the block probably should face the player.

Reply:
so extracting properties ? cause we already extract them, but yea I don't see placement rules that couldn't be derive from some properties.